Add support for latest HTTP Signatures spec draft

ClearlyClaire opened the pull request:
https://www.ietf.org/id/draft-ietf-httpbis-message-signatures-00.html

- rewrite signature params parsing to be much closer to the spec (most notably, do not require parameters to be in quoted strings)
- add support for the `hs2019` signature algorithm (**assumed to be equivalent to RSA-SHA256**, since we do not have a mechanism to specify the algorithm within the key metadata yet)
- add support for (created) and (expires) pseudo-headers and related signature parameters, when using the `hs2019` signature algorithm
- adjust default `headers` parameter while being backwards-compatible with previous implementation
- change the acceptable time window logic from 12 hours surrounding the `Date` header to accepting signatures created up to 1 hour in the future and expiring up to 1 hour in the past (but only allowing expiration dates up to 12 hours after the creation date)
  This doesn't conform with the current draft, as it doesn't permit accounting for clock skew. This, however, should be addressed in a next version of the draft:
  https://github.com/httpwg/http-extensions/pull/1235

Note that for non-hs2019 signatures, this changes the accepted time window from 12 hours surrounding the `Date` header to 1 hour surrounding it. 12 hours was not enough to account for all possible timezone disparities, yet we have never heard of it to be an issue. If it was an issue in practice, it would be rare enough to be unnoticed, and 1 hour should make any issue more obvious. If it never has been an issue, hopefully 1 hour won't be either.

It also adds a bunch of requirements for signatures, those requirements seem to be met by at least Mastodon, Pleroma, PeerTube and Hubzilla.

Error messages should also now be more specific and more informative, should other implementors run into them.

ClearlyClaire committed:
Add support for latest HTTP Signatures spec draft

https://www.ietf.org/id/draft-ietf-httpbis-message-signatures-00.html

- add support for the “hs2019” signature algorithm (assumed to be equivalent
  to RSA-SHA256, since we do not have a mechanism to specify the algorithm
  within the key metadata yet)
- add support for (created) and (expires) pseudo-headers and related
  signature parameters, when using the hs2019 signature algorithm
- adjust default “headers” parameter while being backwards-compatible with
  previous implementation
- change the acceptable time window logic from 12 hours surrounding the “date”
  header to accepting signatures created up to 1 hour in the future and
  expiring up to 1 hour in the past (but only allowing expiration dates up to
  12 hours after the creation date)
  This doesn't conform with the current draft, as it doesn't permit accounting
  for clock skew.
  This, however, should be addressed in a next version of the draft:
  https://github.com/httpwg/http-extensions/pull/1235

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -24,16 +24,27 @@ def signature_verification_failure_code
+      raw_signature    = request.headers['Signature']
+      signature_params = {}
+
+      # TODO: this is really approximative... we should fix it with proper parsing
```
what would the proper parsing look like here? In which way is this implementation approximate? Can you give examples of params where the "proper parsing" would create a hash that differs from the one generated by this code?

ThibG replied:
signature params are defined in the spec as `sig-param    = token BWS "=" BWS ( token / quoted-string )`
in particular, `token` allows more than alphabetic characters, the right-hand side isn't necessarily a quoted-string, and a quoted-string could include a comma, I think

ThibG replied:
I rewrote the parser with `Parslet`, it now correctly handles token values (in addition to quoted strings), as well as commas and escaped quotes in quoted strings, and optional spaces wherever they are allowed. It is not 100% compliant with the spec, though, as my implementation may allow non-ascii or non-printable characters in quoted strings, contrary to the spec, but HTTP's grammar definitions are a pain, and I think this is close enough.

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -24,16 +24,27 @@ def signature_verification_failure_code
-      signature_params[parsed_parts[1]] = parsed_parts[2]
-    end
+  def signature_algorithm
+    signature_params['algorithm'] || 'hs2019'
```
maybe `.fetch('algorithm', 'hs2019')` might be clearer? Up to you

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -45,23 +56,20 @@ def signed_request_account
-    raw_signature.split(',').each do |part|
-      parsed_parts = part.match(/([a-z]+)="([^"]+)"/i)
-      next if parsed_parts.nil? || parsed_parts.size != 3
-      signature_params[parsed_parts[1]] = parsed_parts[2]
```
IIRC there was a reason we wanted to parse it twice. Maybe @Gargron remembers?

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -100,7 +108,8 @@ def request_body
   def verify_signature(account, signature, compare_signed_string)
-    if account.keypair.public_key.verify(OpenSSL::Digest::SHA256.new, signature, compare_signed_string)
+    if (signature_algorithm == 'rsa-sha256' && account.keypair.public_key.verify(OpenSSL::Digest::SHA256.new, signature, compare_signed_string)) ||
+       (signature_algorithm == 'hs2019' && account.keypair.public_key.verify_pss(OpenSSL::Digest::SHA512.new, signature, compare_signed_string))
```
The structure of this code is difficult to read and kind of confusing. Maybe we could extract this into a method? Like this:

```ruby
def signature_matches_digest?(account, signature, compare_signed_string)
  case signature_algorithm
  when 'rsa-sha256'
    account.keypair.public_key.verify(OpenSSL::Digest::SHA256.new, signature, compare_signed_string))
  when 'hs2019'
    account.keypair.public_key.verify_pss(OpenSSL::Digest::SHA512.new, signature, compare_signed_string))
  end
end
```

ThibG replied:
for now I'm going to roll back to just the `verify`, since `hs2019` doesn't necessarily translate to RSASSA-PSS and SHA512, and I couldn't manage to verify examples anyway

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -109,11 +118,15 @@ def verify_signature(account, signature, compare_signed_string)
 
   def build_signed_string(signed_headers)
-    signed_headers = 'date' if signed_headers.blank?
+    signed_headers = (signature_algorithm == 'hs2019' ? '(created)' : 'date') if signed_headers.blank?
```
extract `default_headers` here? or even just `headers`?

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -123,13 +136,36 @@ def build_signed_string(signed_headers)
+    created_time = nil
+    expires_time = nil
+
+    if signature_algorithm == 'hs2019' && signature_params['created'].present?
```
Can we enforce `created` is present for users of the `hs2019` algorithm? That would solve one of the main security issues I've previously had with HTTP-Signatures (Servers can shoot themselves in the foot by not specifying a `created` time)

nightpool replied:
Spec also says we shouldn't verify the signature if `(created)` is part of the `signed_headers` but `created` isn't present in the params but —do we do that?

ThibG replied:
We could enforce the presence of `created`. It's not mandated by the spec, but we are allowed to require more than the spec requires.

As for checking the arguments are here… my code doesn't explicitly fails, but since the arguments are included in the signature input, it will fail anyway. I will change it to fail with an explicit error, though.

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -123,13 +136,36 @@ def build_signed_string(signed_headers)
+    expires_time = nil
+
+    if signature_algorithm == 'hs2019' && signature_params['created'].present?
+      created_time = Time.at(signature_params['created'].to_i).utc
```
we should probably rescue ArgumentError here as well, correct?

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -123,13 +136,36 @@ def build_signed_string(signed_headers)
+    # It is possible for a legitimate user to sign `(created)` but not `(expires)`
+    # therefore, an attacker could set a long expiration time, so set a limit to
+    # how far in the future we allow the expiration to be.
+    expires_time = [expires_time, created_time + 12.hours].min unless created_time.nil?
```
let's add a constant to replace this magic `12.hours`—this should make the code more readable and allow us to get rid of the comment.

nightpool replied:
something like `EXPIRATION_WINDOW_LIMIT = 12.hours`

nightpool replied:
Also, maybe we should only process `expires` if it's part of the signed_headers to avoid allowing the attacker to control this? I can't see how it would be harmful to allow the attacker to control this param, but seems good to do as a matter of best practices anyway. Either way we should limit the signature to at most 12 hours.

ThibG replied:
Addressed the first two comments. Not sure about checking whether `expires` is within the `signed_headers`. I guess with the current implementation, what an attacker could do with it is extend the 1h5min (including clock skew margin) window to 13h (including clock skew margin)… I'm not sure that's worth making the code more complex. The 1h5 minutes is not something mandated by the spec in any way either, the spec doesn't tell what to do if there's no `expires` attribute.

nightpool commented on app/controllers/concerns/signature_verification.rb:
```diff
@@ -109,11 +118,15 @@ def verify_signature(account, signature, compare_signed_string)
     signed_headers.downcase.split(' ').map do |signed_header|
       if signed_header == Request::REQUEST_TARGET
         "#{Request::REQUEST_TARGET}: #{request.method.downcase} #{request.path}"
+      elsif signature_algorithm == 'hs2019' && signed_header == '(created)'
```
spec says to throw an error if the remote server attempts to use `(created)` with the rsa-sha256 algorithm.

ClearlyClaire committed:
Add additional signature requirements

ClearlyClaire committed:
Rewrite signature params parsing using Parslet

ClearlyClaire committed:
Make apparent which signature algorithm Mastodon on verification failure

Mastodon uses RSASSA-PKCS1-v1_5, which is not recommended for new applications,
and new implementers may thus unknowingly use RSASSA-PSS.

ThibG commented:
I've been running that code for a few days, with some extra instrumentation to catch any request that fails to pass signature verification. Here are the issues I've found so far:
- it seems Plume sometimes does not seem to sign `(request-target)`, but it does sign `digest` in its POST requests, so we may relax that requirement to `(request-target)` *or* `digest`
- under some conditions (or maybe only older versions), PeerTube sends requests with an incorrect `Signature` header (https://github.com/Chocobozzz/PeerTube/issues/3075), the previous parsing code did handle that case, though. We could edit the rule to allow for an optional `Signature ` prefix to preserve compatibility, even if it's plain incorrect
- I encountered a handful Mastodon instances with a suspended instance actor (hence #14567), but this isn't a regression, the old code would fail the same way

ClearlyClaire committed:
Add workaround for PeerTube's invalid signature header

The previous parser allowed incorrect Signature headers, such as
those produced by old versions of the `http-signature` node.js package,
and seemingly used by PeerTube.

This commit adds a workaround for that.

ClearlyClaire committed:
Fix `signature_key_id` raising an exception

Previously, parsing failures would result in `signature_key_id` being nil,
but the parser changes made that result in an exception.

This commit changes the `signature_key_id` method to return `nil` in case
of parsing failures.

ClearlyClaire committed:
Move extra HTTP signature helper methods to private methods

ClearlyClaire committed:
Relax (request-target) requirement to (request-target) || digest

This lets requests from Plume work without lowering security significantly.

ThibG commented:
Based on running this on my instance for a few days, I have made a few changes:
- changed the parser to allow an incorrect `"Signature "` prefix to the `Signature` header. This was never in any version of the spec, but this was a result of some libs incorrectly setting to the `Signature` header the value that was defined for the `Authorization` header in a previous version of the draft. Some older PeerTube versions had this issue.
- relaxed the requirement to have `(request-target)` signed to a requirement to have `(request-target)` *or* `digest` signed (`digest` remains mandatory for POST requests), the idea being that one shouldn't be able to reply a GET request as a POST, or a POST with different data, and that is ensured by having this relaxed requirement. This is needed for at least Plume to work correctly.

I have encountered no implementation not signing `Host`, `Date`, or `Digest` for POST requests. Therefore, this PR has the following, stricter requirements, that don't seem to impact backward compatibility with reasonably old fediverse software:
- `host` must be signed (ensures any attack would require the attacker to *intercept* the packet)
- Either `date` or `(created)` must be signed (ensure attacks have a limited time window)
- Either `digest` or `(request-target)` must be signed, and `digest` must be signed whenever a POST request is made (ensures payloads cannot be swapped in attacks)
